crypto/schnorr.c

```diff
@@ -0,0 +1,252 @@
+
+  bn_write_be(Rx, Rxbuf);
+
+  sha256_Init(&ctx);
+  sha256_Update(&ctx, Rxbuf, sizeof(Rxbuf));
+  sha256_Update(&ctx, pub_key, 33);
+  sha256_Update(&ctx, msg_hash, SHA256_DIGEST_LENGTH);
+  sha256_Final(&ctx, digest);
+
+  bn_read_be(digest, e);
+  bn_fast_mod(e, &curve->order);
+  bn_mod(e, &curve->order);
+}
+
+int schnorr_sign(const ecdsa_curve *curve, const uint8_t *priv_key,
+                 const uint8_t *msg, const uint32_t msg_len, uint8_t *sign) {
+  uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
+
+  double_sha256(msg, msg_len, digest);
+
+  return schnorr_sign_digest(curve, priv_key, digest, sign);
+}
+
+int schnorr_sign_digest(const ecdsa_curve *curve, const uint8_t *priv_key,
+                        const uint8_t *digest, uint8_t *sign) {
+  uint8_t pub_key[33] = {0};
+  curve_point R = {0};
+  bignum256 private_key_scalar = {0}, e = {0}, s = {0}, k = {0};
+  rfc6979_state rng = {0};
+
+  ecdsa_get_public_key33(curve, priv_key, pub_key);
+
+  // Compute k
+  init_rfc6979_schnorr(priv_key, digest, &rng);
+  if (generate_k_schnorr(curve, &k, &rng) != 0) {
+    memzero(&k, sizeof(k));
+    return 1;
+  }
+
+  // Compute R = k * G
+  point_multiply(curve, &k, &curve->G, &R);
+
+  // If R.y is not a quadratic residue, negate the nonce
+  bn_cnegate(is_non_quad_residue(&R.y, &curve->prime), &k, &curve->order);
+  bn_mod(&k, &curve->order);
+
+  bn_mod(&R.x, &curve->order);
+  bn_write_be(&R.x, sign);
+
+  // Compute e = H(Rx, pub_key, msg_hash)
+  calc_e(curve, &R.x, pub_key, digest, &e);
+
+  // Compute s = k + e * priv_key
+  bn_copy(&e, &s);
+  bn_read_be(priv_key, &private_key_scalar);
+  bn_multiply(&private_key_scalar, &s, &curve->order);
+  memzero(&private_key_scalar, sizeof(private_key_scalar));
+  bn_addmod(&s, &k, &curve->order);
+  memzero(&k, sizeof(k));
+  bn_fast_mod(&s, &curve->order);
+  bn_mod(&s, &curve->order);
+  bn_write_be(&s, sign + 32);
+
+  return 0;
+}
+
+int schnorr_verify(const ecdsa_curve *curve, const uint8_t *pub_key,
+                   const uint8_t *msg, const uint32_t msg_len,
+                   const uint8_t *sign) {
+  uint8_t digest[SHA256_DIGEST_LENGTH] = {0};
+
+  double_sha256(msg, msg_len, digest);
+
+  return schnorr_verify_digest(curve, pub_key, digest, sign);
+}
+
+int schnorr_verify_digest(const ecdsa_curve *curve, const uint8_t *pub_key,
+                          const uint8_t *digest, const uint8_t *sign) {
+  curve_point P = {0}, sG = {0}, R = {0};
+  bignum256 r = {0}, s = {0}, e = {0};
+
+  bn_read_be(sign, &r);
+  bn_read_be(sign + 32, &s);
+
+  // Signature is invalid if s >= n or r >= p.
+  if (!bn_is_less(&r, &curve->prime) || !bn_is_less(&s, &curve->order)) {
+    return 1;
+  }
+
+  if (!ecdsa_read_pubkey(curve, pub_key, &P)) {
+    return 2;
+  }
+
+  // Compute e
+  calc_e(curve, &r, pub_key, digest, &e);
+
+  // Compute R = sG - eP
+  bn_cnegate(1, &e, &curve->order);
+  bn_mod(&e, &curve->order);
+  point_multiply(curve, &s, &curve->G, &sG);
```

Comment:
Use `scalar_multiply()`, it's specially written to be faster.
```suggestion
  scalar_multiply(curve, &s, &sG);
```